### ISSUE TITLE
Testing Image-To-Text

### ISSUE THREAD

**Issue by ndrean** [dwyl/imgup#131]
I gave Bumblebee a try today. The idea was to provide predictions on image captioning to classify an image so that a user can use/put pre-filled tags to easily filter his images.

It turns out that the predictions are.....not too bad and quite fast., at least locally.

This is supposed to be a car:

<https://dwyl-imgup.s3.eu-west-3.amazonaws.com/40F36F45.webp>

```elixir
Nx.Serving.run(serving, t_img) 
#=>predictions: [
    %{
      label: "beach wagon, station wagon, wagon, estate car, beach waggon, station waggon, waggon",
      score: 0.9203662276268005
    }
  ]
```
Testing with a new query string: `pred=on`to run the model prediction:

```elixir
curl -X GET http://localhost:4000/api?url=https://dwyl-imgup.s3.eu-west-3.amazonaws.com/40F36F45.webp&w=300&pred=on

{"h":205,"w":300,"url":"https://dwyl-imgup.s3.eu-west-3.amazonaws.com/76F195C6.webp","new_size":11642,"predictions":[{"label":"beach wagon, station wagon, wagon, estate car, beach waggon, station waggon, waggon","score":0.9203662276268005}]],"init_size":79294,"w_origin":960,"h_origin":656,"url_origin":"https://dwyl-imgup.s3.eu-west-3.amazonaws.com/40F36F45.webp"}

```

I tested 3 models: "facebook/deit-base-distilled-patch16-224" and "microsoft/resnet-50" and ""google/vit-base-patch16-224".

I don't know if anyone tested it? 

I submit my code in case any reader sees some obvious fault. It runs locally. It is based [on this example](https://github.com/elixir-nx/bumblebee/blob/main/examples/phoenix/image_classification.exs). I did not try to deploy this, [but here is a guide before I forget](https://github.com/elixir-nx/bumblebee/tree/main/examples/phoenix): you need to set up a temp dir.

```elixir
#mix.exs
{:bumblebee, "~> 0.4.2"},
{:nx, "~> 0.6.1"},
{:exla, "~> 0.6.1"},
{:axon, "~> 0.6.0"},
```

I decided to run a GenServer to start the `serving` with the app to load the model, but you can start an `Nx.Serving` in the Aplpication level as well, something like `{Nx.serving, serving: serve(), name: UpImg.Serving}` where the function `Application.serve` defines what is in the GenServer below.

```elixir
defmodule UpImg.GsPredict do
   use GenServer

  def start_link(opts) do
    {:ok, model} = Keyword.fetch(opts, :model)
    GenServer.start_link(__MODULE__, model, name: __MODULE__)
  end

  def serve, do: GenServer.call(__MODULE__, :serving)

  @impl true
  def init(model) do
    {:ok, model, {:continue, :load_model}}
  end

  @impl true
  def handle_continue(:load_model, model) do
    {:ok, resnet} = Bumblebee.load_model({:hf, model})
    {:ok, featurizer} = Bumblebee.load_featurizer({:hf, model})

    {:noreply,
     Bumblebee.Vision.image_classification(resnet, featurizer,
       defn_options: [compiler: EXLA],
       top_k: 1,
       compile: [batch_size: 10]
     )}
  end

  @impl true
  def handle_call(:serving, _from, serving) do
    {:reply, serving, serving}
  end
end
```

and it is started with the app:
```elixir
children = [
  ...,
  {UpImg.GsPredict, [model: System.fetch_env!("MODEL")]}
]
```

The model  - the repo id - is passed as an `env var` so I can very simply change it..

In the API, I use `predict/1` when I upload an image from the browser and run this task in parallel to the S3 upload. It takes a `Vix.Vips.Image`, a transformation of a binary file:

[EDITED]

```elixir
def predict(%Vix.Vips.Image{} = image) do
    serving = UpImg.GsPredict.serve()

    {:ok, %Vix.Tensor{data: data, shape: shape, names: names, type: type}} =
      Vix.Vips.Image.write_to_tensor(image)

    #{width, height, channels} = shape <- wrong, shape should be HWC. Bug corrected.
    t_img = Nx.from_binary(data, type) |> Nx.reshape(shape, names: names)

    Nx.Serving.run(serving, t_img)
    Task.async(fn -> Nx.Serving.run(serving, t_img) end)
  end
```

and use it in the flow:

```elixir
prediction_task = predict(my_image)
...
%{predictions: predictions} = Task.await(prediction_task)
```

**Comment by ndrean** [comment]
Prediction slows down the process, roughly 1.5s per request. I will try to deploy this horror.

The "GET" endpoint - where you pass in an URL of a pic -  works with the query string addition "pred" (no prediction, thus faster is you don't pass one).

curl -X GET http://localhost:4000/api?url=....&w=300&pred=on

The "POST" endpoint - where you submit files from a client via a FormData to the API - also works, but you use a checkbox if you want the prediction (I capture it the same way, via a key "pred", thus there is a constraint on the FormData).

**Comment by ndrean** [comment]
Seems you have better (faster) results when you use `NxServing.batched_run/3` especially when treating several pictures.

The set up is different. `Nx.Serving` as a process is launched in the `Application` level.
```elixir
#Application.ex
children = [
   ...,
    {Nx.Serving, serving: serve(), name: UpImg.Serving, batch_size: 10, batch_timeout: 100}
]

defp serve do
  model = System.fetch_env!("MODEL")
  {:ok, resnet} = Bumblebee.load_model({:hf, model})
  {:ok, featurizer} = Bumblebee.load_featurizer({:hf, model})

  Bumblebee.Vision.image_classification(resnet, featurizer,
      defn_options: [compiler: EXLA],
      top_k: 1,
      compile: [batch_size: 10]
end
```

and then use instead:

```elixir
def predict(%Vix.Vips.Image{} = image) do
    # serving = UpImg.GsPredict.serve()

    {:ok, %Vix.Tensor{data: data, shape: shape, names: names, type: type}} =
      Vix.Vips.Image.write_to_tensor(image)

    {width, height, channels} = shape
   # bug in Vix.Vips, with HWC and WHC....
    t_img = Nx.from_binary(data, type) |> Nx.reshape({height, width, channels}, names: names)

    Task.async(fn -> Nx.Serving.batched_run(UpImg.Serving, t_img) end)
    # Task.async(fn -> Nx.Serving.run(serving, t_img) end)
  end
```

! One must be careful with the async calls. When you run this async task,  say `%Task{} = task = predict(image)`,  you can only get the result back, `Task.await(task)` , from the owner process.

**Comment by ndrean** [comment]
To read an URL and download it with `Finch` using streams (and write it into a temp file) and accept redirects, you can do:

```elixir
{:ok, path} = Plug.upload.random_file("temp-stream")
Finch.build(:get, url)
|> Finch.stream(UpImg.Finch, nil, fn
      {:status, status}, _acc ->
        status

      {:headers, headers}, status ->
        case status do
          302 ->
            Enum.filter(headers, &(elem(&1, 0) == "location")) |> List.first()

          200 ->
            headers

          _ ->
            {:halt, "bad redirection"}
        end

      {:data, data}, headers ->
        case headers do
          {"location", location} ->
            Finch.build(:get, location) |> Api.stream_write(file)

          {:halt, "bad redirection"} ->
            {:error, "bad redirection"}

          _headers ->
            case IO.binwrite(file, data) do
              :ok -> :ok
              {:error, reason} -> {:error, reason}
            end
        end
    end)
```

The memory footprint is low, at the expense of writing the body of the request into a file (but one can just append the chunk in memory if needed).

**Comment by LuchoTurtle** [comment]
Thanks for the excellent write-up, @ndrean , it was super insightful! 
I haven't tried `Bumblebee` but I'm wanting to (when I have more free time). Are you using any specific Hugging Face models in your experiments?

On the topic, you may also find https://github.com/replicate/replicate-elixir as another alternative. Unfortunately, it's tied to their platform, but it might still be fun to tinker with. This is from this AWESOME talk from Charlie Holtz in https://www.youtube.com/watch?v=TfZI5-oQSqI&ab_channel=ElixirConf. It's an awesome video that really highlights how Elixir has great built-in tools to get AI models with LiveView working seamlessly.

**Comment by ndrean** [comment]
Yes. I used [microsoft/resnet model](https://huggingface.co/microsoft/resnet-50). 

Thanks for the "replicate" link. I will give it a try too!

**Comment by ndrean** [comment]
@LuchoTurtle  Thanks! I really enjoyed watching this video, had a lot of fun! 😀
I just realised that you already wrote plenty of good things on this subject before I woke up! So nothing new under the sun for you... 😀 

<https://github.com/dwyl/learn-elixir/issues/212>
<https://github.com/dwyl/imgup/issues/52>

I am just looking at **Image Classification** - namely a weighted list of predictions -  whilst you wanted **Image-to-text**, more ambitious. I just wondered what you would do with the generated text for an image because you need to further process  this response to extract some keys points, if this is what you want.

For example, the **Salesforce/BLIP** is an I2T. I run it in a Livebook, the easiest way to do this. It downloads 1.7Gb... The generated code is:

```elixir
{:ok, model_info} = Bumblebee.load_model({:hf, "Salesforce/blip-image-captioning-base"})

{:ok, featurizer} =
  Bumblebee.load_featurizer({:hf, "Salesforce/blip-image-captioning-base"})

{:ok, tokenizer} =
  Bumblebee.load_tokenizer({:hf, "Salesforce/blip-image-captioning-base"})

{:ok, generation_config} =
  Bumblebee.load_generation_config({:hf, "Salesforce/blip-image-captioning-base"})

generation_config = Bumblebee.configure(generation_config, max_new_tokens: 100)

serving =
  Bumblebee.Vision.image_to_text(model_info, featurizer, tokenizer, generation_config,
    compile: [batch_size: 1],
    defn_options: [compiler: EXLA]
  )

image_input = Kino.Input.image("Image", size: {384, 384})
form = Kino.Control.form([image: image_input], submit: "Run")
frame = Kino.Frame.new()

Kino.listen(form, fn %{data: %{image: image}} ->
  if image do
    Kino.Frame.render(frame, Kino.Text.new("Running..."))

    image =
      image.file_ref
      |> Kino.Input.file_path()
      |> File.read!()
      |> Nx.from_binary(:u8)
      |> Nx.reshape({image.height, image.width, 3})
    %{results: [%{text: text}]} = Nx.Serving.run(serving, image)
    Kino.Frame.render(frame, Kino.Text.new(text))
  end
end)

Kino.Layout.grid([form, frame], boxed: true, gap: 16)
```

I generated an image with Stable Diffusion, and submit it to BLIP. The result is pretty good! 😁, but it is not classification!


<img width="466" alt="Screenshot 2023-10-17 at 19 48 56" src="https://github.com/dwyl/imgup/assets/6793008/0643818f-9dd4-4ecc-8440-5fedf02d5bbc">


This works for me because the model is "deciphered" in some way in Bumblebee. What if I want to use a specific model? That I don't know how to proceed. 

I used a small Image Classification model -300Mb downloaded - embeded into the app as this tends to be much smaller.

Replicate exposes an endpoint. You also need to be careful with the data you submit to get the right balance between speed and accuracy: you might pay too much or pay for nothing if you don't deliver a properly sized image :) 

When you [read this](https://github.com/elixir-nx/bumblebee/blob/main/examples/phoenix/README.md), they naturally stress that the navigator should resize pics instead of a/the server. However, In this git repo, the proposed JS code is a bit ... wordy. 

So down to earth I tried to follow the repo recommendations - at least for a WebApp version -  and I looked at how to do this. In fact, you can get a bunch of resized images from the browser with a `Promise.all` because the browser is efficient  at doing this: a form accepts an image and  you "just" inject it into different resizing canvas that you set, and call `canvas.toBlob`. You can target a thumbnail, or a "ML" sized image (512px) or 3  different sizes to match mobile, pad and full screen 1440px pretty quickly for example. 

One point is the naming: you need a unique base identifier for all these files. It turns out that JS can produce a SHA1 easily, no library, so I used this as a unique naming base, modulo some size identifier. You can also convert into WEBP just like this, and this saves a lot. 
You can upload directly to a bucket, and pass down to the server the 512px file to do the ML stuff. The bucket does his stuff and returns a response back to the client - a bunch of URLs - and the client forwards the responses to the server where you update the socket. Meanwhile, the server did the ML stuff to produce a caption/prediction. It remains to save all this into the DB. With the SHA1 naming, we have a common identifier, almost collision free, so we can update the DB record easily. All easy async client-side and server-side. The main difficulty is the "hook".

The [shinstagram source](https://github.com/cbh123/shinstagram). 
He uses R2 but I did not get the details how he uses the CDN to serve the files.

**Comment by ndrean** [comment]
Yes, I see, a mix. Sometimes I have good predictions, but more often BLIP is superior. For the moment, I don't know.

Thanks for the CanIUseThisModel, I did not know or found. Things are more clear.

Of course, I did not invent the 512px trick, I read it!

The client-code to resize in the browser. This is supposed to be a hook. I tested it partly, it works, but it is not finished.  It is slightly tricky because of processes are all made async in JS to speed u

- I generate a unique name per picture, simply a SHA1, supposed to be unique: `calcSHA1`
- I decided (why not!) to produce 3 versions: a "thumbnail" with max size 200px (I might read this from a dataset, itself populated via an ENV VAR Phoenix side), a "machine-learning" version of target size 512px (width or height), and a pseudo full-screen with max 1440px: `fReader`and `resizeMax``
- I firstly rename (async) all the entries with this SHA1.original_extension
- I process the files (async) through a canvas  to produce 3 new versions in format WEBP (this is just 2 LOC)

This is supposed to be a "hook". It is not finished:
- on submit, I will upload directly to S3 or equivalent, 
- send the 512px version to the Phoenix server to do the ML stuff.
- then merge server-side all the responses and save it into the DB.

```js
<!-- put action="http://localhost:8000"  to check the FormData sent via Necat-->
<html>
  <body>
    <form
      action="http://localhost:4000/api/r2"
      method="POST"
      enctype="multipart/form-data"
    >
      <input type="number" name="w" value="200" />
      <input
        type="file"
        name="files"
        id="img-input"
        accept="image/*"
        multiple
      />
      <input type="checkbox" name="thumb" />
      <input type="checkbox" name="predict" />
      <button type="submit">Submit</button>
    </form>
    <p id="response"></p>

    <script>
      // code taken from MDN
      async function calcSHA1(file) {
        const arrayBuffer = await file.arrayBuffer();
        const hash = await window.crypto.subtle.digest("SHA-1", arrayBuffer);
        const hashArray = Array.from(new Uint8Array(hash));
        const hashHex = hashArray
          .map((b) => b.toString(16).padStart(2, "0"))
          .join("");
        return hashHex;
      }

      function resizeMax(w, h, MAX) {
        const MAX_WIDTH = MAX;
        const MAX_HEIGHT = MAX;

        if (w > h) {
          if (w > MAX_WIDTH) {
            h = h * (MAX_WIDTH / w);
            w = MAX_WIDTH;
          }
        } else {
          if (h > MAX_HEIGHT) {
            w = w * (MAX_HEIGHT / h);
            h = MAX_HEIGHT;
          }
        }
        return { w, h };
      }

      let fList = [];

     document
      .querySelector("#img-input")
      .addEventListener("change", async (e) => {
        try {
          renamedFiles = await Promise.all(
            [...e.target.files].map(async (file) => {
              const ext = file.type.split("/").at(-1);
              const SHA1name = await calcSHA1(file);
              return new File([file], `${SHA1name}.${ext}`, {
                type: file.type,
              });
            })
          );

          fList = await Promise.all(
            renamedFiles.map((file) => processFile(file))
          );

          // preview
          fList.forEach(([_ml, preview, _origin]) => {
            const newDataUrl = URL.createObjectURL(preview);
            const img = document.createElement("img");
            img.src = newDataUrl;
            document.body.appendChild(img);
          });

          // build the final "pre-list"
          fList = [...renamedFiles, fList.flat()].flat();
        } catch (err) {
          throw new Error(err);
        }
      });

      // for each file, produce 3 versions in WEBP
      async function processFile(file) {
        const files = await Promise.all([
          fReader(file, 512),
          fReader(file, 200),
          fReader(file, 1440),
        ]);
        return files;
      }

      function fReader(file, MAX) {
        return new Promise((resolve, reject) => {
          if (file) {
           // pass a data:url to an Image
            const img = new Image();
            img.src = URL.createObjectURL(file);

            img.onload = function () {
              // calculate the sizes depending on the MAX constraint
              const { w, h } = resizeMax(img.width, img.height, MAX);

              const canvas = document.createElement("canvas");
              if (canvas.getContext) {
                const ctx = canvas.getContext("2d");
                canvas.width = w;
                canvas.height = h;
                ctx.drawImage(img, 0, 0, w, h);

                // async convert the image from the canvas into a Blob and convert into WEBP format
                canvas.toBlob(
                  (blob) => {
                    const name = file.name.split(".")[0];
                    const convertedFile = new File(
                      [blob],
                      `${name}-m${MAX}.webp`,
                      {
                        type: "image/webp",
                      }
                    );
                    resolve(convertedFile);
                  },
                  "image/webp",
                  0.75
                );
              }
            };
            img.onerror = function () {
              reject("Error loading image");
            };
          } else {
            reject("No file selected");
          }
        });
      }

      const form = ({ action, method } = document.forms[0]);

      form.onsubmit = async (e) => {
        e.preventDefault();
        let fd = new FormData(form);
        fd.delete("files");
        fList.forEach((file) => fd.append("files", file));

        try {
         // for the moment, submit everything to the Phoenix backend to check
          const req = await fetch(action, { method, body: fd });

          if (req.ok) {
            const response = await req.json();
            return setResponse(form, response);
          } else {
            throw new Error("File upload failed");
          }
        } catch (error) {
          throw new Error(`Error uploading file:", ${error}`);
        }
      };

      // for the moment, print in the browser. Future: send the supposed received URLs from bucket back to the backend.
      function setResponse(form, response) {
        preview.innerHTML = "";
        form.reset();
        document.querySelector("#response").innerText =
          JSON.stringify(response);
      }
    </script>
  </body>
</html>
```

**Comment by LuchoTurtle** [comment]
@ndrean 
```elixir
    t_img = Nx.from_binary(data, type) |> Nx.reshape({height, width, channels}, names: names)
```

I'm having trouble with this part. I keep stumbling upon this error when trying to reshape the tensor so I can feed it into the `resnet-50` model.

```sh
** (ArgumentError) cannot reshape, current shape {11708} is not compatible with new shape {224, 224, 3}
```

I know **for sure** that the image is resized according to the model's specification (`224x224`) up until this point.
I don't know what I'm doing wrong, I'm trying to follow `Bumblebee`'s guide to image classification.

**Comment by ndrean** [comment]
@LuchoTurtle  ah yes I remember now, I had this too, it was bug, and my code above was good with the bug until the maintainer corrected it, so its wrong now. but I did not correct it above... 

The correct shape is a HWC tuple: width and height were inverted , you see what I did?


Make sure to have the latest version.  I think this should work.

```elixir
 {:ok, %Vix.Tensor{data: data, shape: shape, names: names, type: type}} =
          Image.write_to_tensor(image)

        t_img = Nx.from_binary(data, type) |> Nx.reshape(shape, names: names)

        Nx.Serving.batched_run(UpImg.Serving, t_img)
```

FYI you cannot deploy this on a small machine, you probably need 1GB RAM. I will probably come back to this as I want to finish this little project.

**Comment by nelsonic** [comment]
A 2GB RAM VPS instance on OVH IS €3.50/month —>  https://github.com/dwyl/learn-devops/issues/64 💭

**Comment by ndrean** [comment]
About "prompt engineering": 
<https://prmpts.ai/blog/what-is-prompt-engineering>

<img width="1187" alt="Screenshot 2023-11-06 at 20 00 14" src="https://github.com/dwyl/imgup/assets/6793008/8c5f7ff3-46eb-4760-ae1b-274f2bb89a04">

**Comment by LuchoTurtle** [comment]
[I argue that "prompt engineering" is engineering at all 😅](https://www.reddit.com/r/ChatGPT/comments/12wxn1w/prompt_engineering_is_easy_as_shit_and_anybody/).

But I do understand that there's an art to it. Refining models' output to get what you want is not easy, per se, but rather a matter of trial and error and specificity. It's definitely a skill but I honestly can't see the "engineering" part of it - it can be boiled down to clarity in communication and having to work with some quirks that `GPT` or any other `LLMs` may have. But hey, maybe I'm an idiot and I'm spewing nonsense, I don't know 😅.

Although, I have to admit, I've tried `Stable Diffusion` much for generative art than `LLMs` (though I'm keen on biting the bullet and paying 20 quid a month for access to `OpenAI`'s API after their dev day at https://openai.com/blog/new-models-and-developer-products-announced-at-devday).

From what I've tried, I think "prompt engineering" is **much harder** with diffusion models than LLMs. But even then, you can circumvent issues with [inpainting](https://stable-diffusion-art.com/inpainting_basics/) and [`ControlNet`](https://github.com/lllyasviel/ControlNet) to get more accurate results rather easily (though it's still very much trial and error, you can't ever get *exactly* what you want, just what's the *closest to what you want*).

For example, what I found to generate cool Ghibli-style images with `Stable Diffusion`, I found that I had to work **much more** than simply using `ChatGPT` or any other `LLM`.

- having to try different models and checkpoints from `Civit.AI` to see the best results (for example, [`EasyFluff`](https://huggingface.co/zatochu/EasyFluff/tree/main)).
- knowing which `VAE` to use to get your preferred results (for example https://civitai.com/models/23906/kl-f8-anime2-vae).
- reducing noise and improving consistency with extensions like https://github.com/Seshelle/CFG_Rescale_webui.
- trying different embeddings for [negative prompts](https://huggingface.co/datasets/Nerfgun3/bad_prompt) or positive prompts.
- knowing which [`LORA`](https://stable-diffusion-art.com/lora/) to use (like https://civitai.com/models/82098/add-more-details-detail-enhancer-tweaker-lora)

This by itself is much more work to just yield fair results with generative art, something that is much more streamlined with `LLMs` (or downright not present) and prompting.
I found prompting in diffusion models absolutely chaotic. But I've seen patterns, and I've had luck trying to follow imageboard tags and I assume many models are trained with these imageboards in mind, because they perform much better when I use these tags.

For example, I tend to follow a pattern for positive prompts

```
establish style +  number of characters +  the camera and/or landscape and/or scene properties and using `"BREAK"` between different subjects that I want in the pictures (to prevent getting mixed up).
```

Adding weight to each tag and you can go from there.

## Is this engineering?

So is this workflow engineering? I don't believe so. It's not deterministic by nature. It's just proper concise communication. It's a skill, but I don't think there's anything esoteric about it.

I liked this answer from https://news.ycombinator.com/item?id=36971327.

```
> Is Prompt Engineering a Thing?

Yes, it's a dumb name for the skill of modifying your prompts and questions to the LLM in a way that produces better results than if you just asked for what you wanted plainly. As language models get better, this might become obsolete.

> I'm trying to research the subject but I don't see much evidence that companies are racing to hire prompt engineers.

Because it's not really a job. Think of it like using the Google search engine - being able to search well is something you can get better at but being a "Google search-er" isn't a career or a job you'll see openings for.
```

All in all, aside from my obvious ramble and digression, it's still an interesting read @ndrean . Because although I don't think it's engineering, it's a highly valuable skill that I want to get better at!

**Comment by ndrean** [comment]
Nice @LuchoTurtle , you look pretty advanced!

Do you use only a Livebook to test all this?

There is indeed some vocabulary to ingest to enter this world. Being able to name things that are really useful is a powerful skill 😀 but feels sometimes like much ado about "almost" nothing.

Playing with images gives an immediate wow effect. I highly recommend <https://github.com/cbh123/emoji> by the same guy who did Shinstagram. 

I still have basic questions: how do you use these tools in practice to run this on production? Is it "embedded"? downloaded and stored on disk...Is there an endpoint to which you submit data? Costs? etc
 
I do more modest down-to-earth things, more on the LLM side. My first step was image captioning. For example, to run this in practice, I download the data on a server and mount bind into the running container. This is not totally straightforward as I don't want to run a 2.5Gb image.

Another barrier is that few models can be used by the Elixir eco-system. I finally found something:  

<https://twitter.com/sean_moriarity/status/1715758666001928613>

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">An explanation on how we add models to Bumblebee (<a href="https://twitter.com/toranb?ref_src=twsrc%5Etfw">@toranb</a> asked on EEF slack and I thought it would be a good write up here).<br><br>The first thing to note is that almost all of the models have significant overlap in implementation details. A transformer is a transformer. There are…</p>&mdash; Sean Moriarity (@sean_moriarity) <a href="https://twitter.com/sean_moriarity/status/1715758666001928613?ref_src=twsrc%5Etfw">October 21, 2023</a></blockquote> 


Lastly, another barrier IMO is LiveView. Compared to [Streamlit](https://streamlit.io/), it is far behind. Liveview is still complicated and fragile: navigation, "liveview session" is obscur. I had some errors I still don't understand.  For example, I used a separate "html.heex" file that for some reason gave me double renderings.  When I put the same markup into the `render` function of the liveview, it worked. I also have some cache problems: you change the code but it doesn't render. Few headaches...

**Comment by ndrean** [comment]
You can spend your life just watching youtube. However, this is worth watching, you learn something: running ML in the browser, VERY instructive. This helps you to understand step by step this Huggingface world and consequently puts some light on the Elixir Bumblebee world (because honestly, they don't help you 😏).

<https://www.youtube.com/watch?v=QdDoFfkVkcw>


![Screenshot 2023-11-11 at 10 54 06](https://github.com/dwyl/imgup/assets/6793008/0df61f2d-b438-44e3-9b46-b464fc5d2336)

**Comment by nelsonic** [comment]
Very good video. Thanks for sharing @ndrean 
Please have a read of: https://github.com/dwyl/image-classifier and share your thoughts. 🙏 

As for the job/title of "Prompt Engineer" ... 
While it's super "hot" right now to know how to refine queries to get pre-trained models to give useful results ...

![image](https://github.com/dwyl/imgup/assets/194400/faa739db-1434-4df7-8513-d6359b2d24fb)

I cannot help but think that this is something a 5-year-old child can do quite effectively.
So it's only a matter of time before the "Prompt Engineers" are replaced.

<img width="456" alt="image" src="https://github.com/dwyl/imgup/assets/194400/0df79594-b8ad-432a-9b21-dbea490c4b14">

What might not be replaced as quickly - though will eventually - are specific subject-matter-experts who use the corpus of knowledge to answer specific questions that non-experts wouldn't even think of. 💭 
But I honestly think as all knowledge gets sucked into ever more powerful LLMs 
_and_ the LLMs have all the questions and _answers_ they will be able to auto-suggest the prompts.
So even a child will be able to prompt their way into a Nobel Prize. 😉

**Comment by ndrean** [comment]
@nelsonic 
I looked quickly into the <https://github.com/dwyl/image-classifier> repo.  Looks good. A few remarks.

1) Are you able to run this on Fly.io?  Because I see that your Dockerfile uses the standard user "nobody" but Hugginface recommends a user 1000. <https://huggingface.co/spaces/jonatanklosko/chai/blob/main/Dockerfile>. This repo can be a reference: <https://huggingface.co/docs/hub/spaces-sdks-docker#permissions>. However, it downloads the model during the build stage, and I found this complicated. You opted to copy the model data from your host into the image <https://github.com/dwyl/image-classifier/blob/d7205ca4a97a1d582436d5cc9d781eb80b6311b2/Dockerfile#L56>, but you don't use ` ENV BUMBLEBEE_OFFLINE=true` in the Dockerfile. I believe that it will download the image, wouldn't it? I believe your image  should use a volume to grab the data and contain only the running code. But if it works this way (the model is small?), then why not, it is not meant to be scaled I presume. Another detail is that the .bumblebee data is also persisted in the Github repo, but shouldn't it be in an LFS? or not at all.


2) You pass a base64 string to render the resized image, but why do you use a form to wrap the `img` tag? <https://github.com/dwyl/image-classifier/blob/d7205ca4a97a1d582436d5cc9d781eb80b6311b2/lib/app_web/live/page_live.html.heex#L22>


3) Why do you need this [pre-process-image](https://github.com/dwyl/image-classifier/blob/d7205ca4a97a1d582436d5cc9d781eb80b6311b2/lib/app_web/live/page_live.ex#L75)? 

4) Shouldn't the async task be `async_nolink`, because if the serving fails, you may not want the main process to get killed.

5) You also have the library [stb_image](https://hexdocs.pm/stb_image/StbImage.html) instaed of Vix. This can further reduce the image size. [An example](https://github.com/bcat-eu/jimmy).

**Comment by nelsonic** [comment]
@ndrean Great feedback as always. CC: @LuchoTurtle (who is currently working on the Fly.io deployment/update ...)

**Comment by ndrean** [comment]
ah ok, didn't look at who did it. So with Lucho, its in good hands :) I am interested to see your result as I want to deploy some thing similar but on a VPS (but using a bucket to save the images and a DB to save the list of images/captions per user). Nothing huge but not obvious :)

**Comment by ndrean** [comment]
Fly.io [volumes](https://fly.io/docs/apps/volume-manage/)

I would try to copy the .bumblebee data you downloaded via Bumblebee into a fly volume. I think this can be done in the fly.toml with (not totally sure):

 [mounts] source=.bumblebee destination=/my-volume

Then you can get rid of the .bumblebee copy command in both stages, and reference the new location in the runner stage with:

ENV BUMBLEBEE_CACHE_DIR=/my-volume ?
ENV BUMBLEBEE_OFFLINE=true

However, not sure your image will fit in a 256MB machine....

**Comment by LuchoTurtle** [comment]
Thanks for the feedback @ndrean , always appreciated!
By the way, thank you for the video! Watched it all the way through, and it was immensely useful! 

1 - Thanks, I didn't know about the "nobody" user had any impact. Will change it :)
And yes, I was trying to cache the model and was hoping to do this all on `fly.io`. Meaning that on the first execution of the app in `fly.io`, the model would be downloaded into `.bumblebee` (hence why I created this directory) and then on subsequent runs, `LiveView` would fetch the local model from it. I thought setting `BUMBLEBEE_OFFLINE` was optional (I thought it was a flag to ALWAYS fetch locally) because I was under the impression that by setting the `CACHE_DIR`, it would use the local model. Apparently, it doesn't, hence why I'm trying to fix it.

2 - I wrap the `<img>` with a form so the user can click on the image again and upload another image if they want to.

3 - I was having trouble with the tensor dimensions initially. Because models usually work on a specific colourspace and without alpha (it's data that is not relevant), I wrote that little function that can be used anywhere. If flattens the alpha out, converts the colourspace and formats/reshapes the tensor to the correct format. That's how I got this to work :p 

4 - OOh, interesting! Thank you for the suggestion :) 


Regarding using volumes, I'm tempted to do so. I want to first try to get the model during the build stage (as you've mentioned)

**Comment by ndrean** [comment]
Another interesting repo to prepare yourself to lose your job??

<https://github.com/KillianLucas/open-interpreter/>

![Screenshot 2023-11-13 at 15 06 49](https://github.com/dwyl/imgup/assets/6793008/f05550d1-a2e3-4b2c-9de0-df260687787e)

**Comment by LuchoTurtle** [comment]
> Another interesting repo to prepare yourself to lose your job??
> 
> [KillianLucas/open-interpreter](https://github.com/KillianLucas/open-interpreter/)
> 
> ![Screenshot 2023-11-13 at 15 06 49](https://user-images.githubusercontent.com/6793008/282483289-f05550d1-a2e3-4b2c-9de0-df260687787e.png)

Looks like https://github.com/Significant-Gravitas/AutoGPT :P

**Comment by LuchoTurtle** [comment]
> No it won't download the model in the build stage unless we explicitly "pre-run" the `Bumblebee.load` and friends, in some `mix` command. Take a look at "Livebeats with whispers" at how they do it the [Dockerfile](https://github.com/fly-apps/live_beats/blob/57f5c0f14277c6e476b9998c953e58d7deb0cd9d/Dockerfile#L61): they are explicit. But this becomes intricate and I don't like this way to do: the model should be in a separate volume, and passed via an env var.

Thank you for the reply. I was trying to get it to work with something similar to that. I want to give both options a whirl but I'm having trouble with actually getting my Dockerfile to work by running something like

```dockerfile
RUN /app/bin/app eval 'App.Application.serving()'
```

But it's not working.

Trying to debug locally but even then, it's a pain and even dumping logs in intermediate Docker layers isn't allowing me to see the filesystem at each step of the build stage.

I see your POV, though. Having it in the dockerfile makes it too tightly coupled but I'm still wanting to give it a try to document both approaches 👌

**Comment by ndrean** [comment]
I hate this "doesn't work for me", but here we are. Same for me, doesn't work because if I recall correctly,  it says "can't find "/app/bin/app".

When I run a release version, `Application.serving()`  works, but putting this in the Dockerfile (which minics what we do by hand, no?), well, doesn't....I did not find an answer.

**Comment by ndrean** [comment]
Mine

```
ARG ELIXIR_VERSION=1.15.5
ARG OTP_VERSION=26.0.2
ARG DEBIAN_VERSION=bullseye-20230612-slim

ARG BUILDER_IMAGE="hexpm/elixir:${ELIXIR_VERSION}-erlang-${OTP_VERSION}-debian-${DEBIAN_VERSION}"
ARG RUNNER_IMAGE="debian:${DEBIAN_VERSION}"

FROM ${BUILDER_IMAGE} as builder

ARG MIX_ENV
# install build dependencies
RUN apt-get update -y && apt-get install -y build-essential git libmagic-dev curl\
  && apt-get clean && rm -f /var/lib/apt/lists/*_*

WORKDIR /app

RUN mix local.hex --force && \
  mix local.rebar --force

ENV MIX_ENV="prod"

COPY mix.exs mix.lock ./
RUN mix deps.get --only $MIX_ENV
RUN mkdir config

COPY config/config.exs config/${MIX_ENV}.exs config/
RUN mix deps.compile

COPY priv priv
COPY assets assets
COPY lib lib
RUN mix assets.deploy

RUN mix compile

COPY config/runtime.exs config/
COPY rel rel
RUN mix release

################################

FROM ${RUNNER_IMAGE}

ARG MIX_ENV

RUN apt-get update -y && apt-get install -y libstdc++6 openssl libncurses5 locales libmagic-dev \
  && apt-get clean && rm -f /var/lib/apt/lists/*_*

RUN sed -i '/en_US.UTF-8/s/^# //g' /etc/locale.gen && locale-gen

ENV LANG en_US.UTF-8
ENV LANGUAGE en_US:en
ENV LC_ALL en_US.UTF-8

WORKDIR "/app"
RUN chown nobody /app

ENV MIX_ENV="prod"

ENV BUMBLEBEE_CACHE_DIR=/app/bin/.bumblebee/blip
ENV BUMBLEBEE_OFFLINE=true

COPY --from=builder --chown=nobody:root /app/_build/${MIX_ENV}/rel/up_img ./

USER nobody

EXPOSE 4000

CMD ["/app/bin/server"]
```

**Comment by LuchoTurtle** [comment]
Yep, precisely the issue I'm getting. The funny thing is that if I run `RUN mix run -e 'App.Application.serving()' --no-start --no-halt`, it clearly downloads the model but fails afterwards.

![image](https://github.com/dwyl/imgup/assets/17494745/bfec4d8f-644a-426e-8756-5872fe174b9c)

![image](https://github.com/dwyl/imgup/assets/17494745/db8a2a6c-1920-4e23-a3e2-b777976e1391)

**Comment by ndrean** [comment]
Yes. By the way, are you running this locally?

**Comment by ndrean** [comment]
MY Docker commands to vercome this: a volume!

1) I create a volume (yes a bit intricate to do...)

```bash
docker run --env-file .env-docker -p 4000:4000 -it --rm --name up-img-cont -v blip:/app/bin/.bumblebee/blip/:ro up-img
```

2) Run the image, inject env and bind mount the volume

```bash
docker run --env-file .env-docker -p 4000:4000 -it --rm --name up-img-cont -v blip:/app/bin/.bumblebee/blip/:ro up-img
```

**Comment by LuchoTurtle** [comment]
Yeah, I'm building the docker locally to surpass errors. 
To be honest, if we're running this failling line *just to get the model*, we only care that it downloads. So if we do something like:

```dockerfile
RUN mix run -e 'App.Application.serving()' --no-start --no-halt; exit 0
```

It should work.

And thanks for the "Volume approach". Once I get this one working, I'll do that one for sure (I really want to document both options)! (though, I ought to admit, by "Volumes", I thought you meant **Fly.io Volumes**, not `Docker` ones).

I also have a question, since I'm really not well-versed with Docker. I understand that those commands work fine locally. But if I wanted to deploy it to `Fly.io`, I'd have to somehow create volumes and bind that on the `Dockerfile`, correct? Do you have any reference for those? I'd google for it but if you already know the answer, perhaps you can point me to the right direction :)

**Comment by ndrean** [comment]
You maybe right with your "exit 0", but I even tried to run only the `Bumblebee.load` and it failed, so I gave up and wnet ot use volumes.

Yes, in the previous post, I mentionned the Fly.io [volumes](https://fly.io/docs/apps/volume-manage/). You don't run docker command in fly, but use a "fly.toml".  


Some thing like [mounts] source=$(pwd)/.bumblebee destination=/my-volume

**Comment by ndrean** [comment]
Because I don't know how to ssh into a fly machine. I mean, I don't don't which machine nor its ssh port nor ip address,  so I can't 

```bash
scp -P 22 -r .bumblebee/ <fly-machine>@<fly-ip>:your-volume
```

**Comment by ndrean** [comment]
I tried again by curiosity. You will probably encounter this error. This is curious because we have "no-start".

```
 => ERROR [builder 15/18] RUN  mix run -e  'UpImg.Ml.serve()' --no-start                                                                                                                                                               30.9s
------
 > [builder 15/18] RUN  mix run -e  'UpImg.Ml.serve()' --no-start:
|====================
13.08 [output clipped, log limit 2MiB reached]
30.45 ** (exit) exited in: GenServer.call(EXLA.Client, {:client, :host, [platform: :host]}, :infinity)
30.45     ** (EXIT) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started
```

**Comment by LuchoTurtle** [comment]
Precisely. In fact, even though I got this error while running https://github.com/dwyl/imgup/issues/131#issuecomment-1808265459, I could circumvent it by adding `; exit 0`.

But now, for some odd reason, I keep getting this error and it's not downloading the models anymore as it was before. I've tried reverting the code i had when I commented https://github.com/dwyl/imgup/issues/131#issuecomment-1808265459 but it's not working anymore. 

?? weird

**Comment by ndrean** [comment]
I tried with success 
```bash
RUN  mix run -e  'UpImg.Ml.serve()' --no-start; exit 0
```

**Comment by LuchoTurtle** [comment]
That works, for sure. But it's maybe not downloading the model. When I commented in https://github.com/dwyl/imgup/issues/131#issuecomment-1808265459, I could see the progress of the model being downloaded. 

But not anymore 🫥, even though the code is exactly the same.

But if you check your container and see the model there, great news 🥳

**Comment by ndrean** [comment]
Ah I did not copy in the last stage

**Comment by ndrean** [comment]
Well it fails since nothing is downloaded indeed

**Comment by ndrean** [comment]
The only way is to copy your local folder into the image and forget about this download. I don't understand how Livebeats made it work.

**Comment by LuchoTurtle** [comment]
If you run

```dockerfile
RUN mix run -e 'App.Application.serving()' --no-start --no-halt; exit 0
COPY .bumblebee/ .bumblebee
```

You need to make sure that `ENV BUMBLEBEE_OFFLINE=true` is **not enable** before this step, otherwise it will not look for the repo. 

For example, here's my dockerfile.

```dockerfile
# Find eligible builder and runner images on Docker Hub. We use Ubuntu/Debian
# instead of Alpine to avoid DNS resolution issues in production.
#
# https://hub.docker.com/r/hexpm/elixir/tags?page=1&name=ubuntu
# https://hub.docker.com/_/ubuntu?tab=tags
#
# This file is based on these images:
#
#   - https://hub.docker.com/r/hexpm/elixir/tags - for the build image
#   - https://hub.docker.com/_/debian?tab=tags&page=1&name=bullseye-20231009-slim - for the release image
#   - https://pkgs.org/ - resource for finding needed packages
#   - Ex: hexpm/elixir:1.15.7-erlang-26.0.2-debian-bullseye-20231009-slim
#
ARG ELIXIR_VERSION=1.15.7
ARG OTP_VERSION=26.0.2
ARG DEBIAN_VERSION=bullseye-20231009-slim

ARG BUILDER_IMAGE="hexpm/elixir:${ELIXIR_VERSION}-erlang-${OTP_VERSION}-debian-${DEBIAN_VERSION}"
ARG RUNNER_IMAGE="debian:${DEBIAN_VERSION}"

FROM ${BUILDER_IMAGE} as builder

# install build dependencies (and curl for EXLA)
RUN apt-get update -y && apt-get install -y build-essential git curl \
    && apt-get clean && rm -f /var/lib/apt/lists/*_*

# prepare build dir
WORKDIR /app

# install hex + rebar
RUN mix local.hex --force && \
    mix local.rebar --force

# set build ENV
ENV MIX_ENV="prod"
ENV BUMBLEBEE_CACHE_DIR="/app/.bumblebee"

# install mix dependencies
COPY mix.exs mix.lock ./
RUN mix deps.get --only $MIX_ENV
RUN mkdir config

# copy compile-time config files before we compile dependencies
# to ensure any relevant config change will trigger the dependencies
# to be re-compiled.
COPY config/config.exs config/${MIX_ENV}.exs config/
RUN mix deps.compile

COPY priv priv

COPY lib lib

COPY assets assets

COPY .bumblebee/ .bumblebee

# compile assets
RUN mix assets.deploy

# Compile the release
RUN mix compile

# IMPORTANT: This downloads the HuggingFace models from the `serving` function in the `lib/app/application.ex` file. 
# And copies to `.bumblebee`.
RUN mix run -e 'App.Application.load_models()' --no-start --no-halt; exit 0
COPY .bumblebee/ .bumblebee

# Changes to config/runtime.exs don't require recompiling the code
COPY config/runtime.exs config/

COPY rel rel
RUN mix release

# start a new build stage so that the final image will only contain
# the compiled release and other runtime necessities
FROM ${RUNNER_IMAGE}

RUN apt-get update -y && \
  apt-get install -y libstdc++6 openssl libncurses5 locales ca-certificates \
  && apt-get clean && rm -f /var/lib/apt/lists/*_*

# Set the locale
RUN sed -i '/en_US.UTF-8/s/^# //g' /etc/locale.gen && locale-gen

ENV LANG en_US.UTF-8
ENV LANGUAGE en_US:en
ENV LC_ALL en_US.UTF-8

WORKDIR "/app"
RUN chown nobody /app

# set runner ENV
ENV MIX_ENV="prod"
ENV EXS_DRY_RUN="true"
ENV BUMBLEBEE_CACHE_DIR=/app/.bumblebee
ENV BUMBLEBEE_OFFLINE=true

# Adding this so model can be downloaded
RUN mkdir -p /nonexistent

# Only copy the final release from the build stage
COPY --from=builder --chown=nobody:root /app/_build/${MIX_ENV}/rel/app ./
COPY --from=builder --chown=nobody:root /app/.bumblebee/ ./.bumblebee

USER nobody

# If using an environment that doesn't automatically reap zombie processes, it is
# advised to add an init process such as tini via `apt-get install`
# above and adding an entrypoint. See https://github.com/krallin/tini for details
# ENTRYPOINT ["/tini", "--"]

# Set the runtime ENV
ENV ECTO_IPV6="true"
ENV ERL_AFLAGS="-proto_dist inet6_tcp"
ENV BUMBLEBEE_CACHE_DIR=/app/.bumblebee
ENV BUMBLEBEE_OFFLINE=true

CMD ["/app/bin/server"]
```

It runs successfully.

When I run the container instance locally, I can inclusively see the models downloaded in `.bumblebee`.

<img width="1007" alt="image" src="https://github.com/dwyl/imgup/assets/17494745/f2c8407e-c3ec-4e1b-aa5e-0656f66aae4b">

Though the naming of these files is weird. It's just mambo jambo. I'll try to load these models locally.

**Comment by ndrean** [comment]
ah yes, I had already loaded some ENv vars to run `mix phx.server`, so  Docker picked up these env vars. 
To set `BUMBLEBEE_OFFLINE=false` in the Dockerfile to be sure this won't happen again

**Comment by ndrean** [comment]
My image inflated from 500Mb to 2.4Gb!.

Look at the running container: 2.4Gb memory! The peak is when I uploaded an image and ran the Image-To-Text serving. Not sure all this is super performant.

<img width="1005" alt="Screenshot 2023-11-13 at 16 58 31" src="https://github.com/dwyl/imgup/assets/6793008/e4c7cd88-4436-4f50-8ff9-960d33e1f6ba">

**Comment by LuchoTurtle** [comment]
That is to be expected, though.
If you check https://fly.io/phoenix-files/speed-up-your-boot-times-with-this-one-dockerfile-trick/, it is mentioned that's the only downsize - your docker image will be much bigger, since it quite literally has your model downloaded there.

<img width="883" alt="Screenshot 2023-11-13 at 16 15 00" src="https://github.com/dwyl/imgup/assets/17494745/9623bdc7-8eec-4ae3-8083-de86d1bff3a4">

However, I'm having trouble actually *using the local model* on runtime. I know I should use `{:local, `path/to/model`}` (according to https://hexdocs.pm/bumblebee/Bumblebee.html#t:repository/0). But, as I've shown, the downloaded model has the file names randomized :/

**Comment by ndrean** [comment]
Yes, that was one of the reasons I mentioned. The difference is that you can run several instances at a minimal cost. You would run 2.5Gb x 2 =5Gb whilst with a volume, it will run 2Gb + 0.5Gb * 2 = 3Gb

**Comment by ndrean** [comment]
One thing I am sure it works is (sorry I am not selling the volume but I want to summarise):
- I run `{:ok, model_info} = Bumblebee.load_model({:hf, model}` in the Elixir code in the `serve/0` function
- use a volume "blip"
- do nothing in the Dockerfile
- pass an env var BUMBLEBEE_CACHE_DIR=/app/bin/.bumblebee/blip (in my case) and mount bind this volume "blip" to this env var

**Comment by ndrean** [comment]
The path to local: I guess it must be absolute, no?

**Comment by LuchoTurtle** [comment]
Nah, it's okay, selling the volume absolutely makes sense 😂 . It's the best option to save resources and to scale horizontally.

Though I don't understand, if you are using `{:ok, model_info} = Bumblebee.load_model({:hf, model}`, aren't you downloading the model? Unless you set the `:local` property to a path (and with this, [Bumblebee with look for the cache directory](https://hexdocs.pm/bumblebee/Bumblebee.html#t:repository/0)), the model is downloading from HuggingFace every time, no?

The path should be absolute, yes.

**Comment by ndrean** [comment]
In the build stage, I do not invoke any RUN mix..., so it is just compiling the code. In fact, I never run the code.

When I run the image, then, yes, it is looking for the cache if nay. If you set the env BUMBLEBEE_CACHE_DIR, it will look for it. If you set `{:local, path}`, it will probbly look for it there. This I am testing

**Comment by ndrean** [comment]
Does not seem so: <https://github.com/elixir-nx/bumblebee/blob/57bdcced8d0df356093787ef38c02d76230b72a2/lib/bumblebee/huggingface/hub.ex#L52>

**Comment by LuchoTurtle** [comment]
Hm, interesting.

This repo's `fly.io` page is set to suspend itself every time there's no activity in one hour. I just accessed it and it was inactive.

<img width="1206" alt="image" src="https://github.com/dwyl/imgup/assets/17494745/d8122575-8d82-4e02-8723-e1823475cb17">


It re-downloaded `ResNet-50`. 🤔

If it's truly cached like you showed, what's the point of https://fly.io/phoenix-files/speed-up-your-boot-times-with-this-one-dockerfile-trick/#bumblebee?

**Comment by ndrean** [comment]
Your machine must be destroyed after one hour? Restart from zero?

Then as we said, their code doesn't work, or we don't see what is wrong

**Comment by ndrean** [comment]
When I use `{:local, "/app/bin/.bumblebee/blip"}`, I get an error "no config file found in the given repository."

However, when I inspect the running container with a `docker exec -it test bash` and `ls`, I find  the bind mounted folder and its populated.

Then when I run serve with :local and an absolute hard-coded path, same error. When I use `:hf`with a populated folder from a previous download, it works.

**Comment by LuchoTurtle** [comment]
Yeah. I have a feeling that `:local` is meant to be used when you directly download the model and add it to the repo manually.

**Comment by LuchoTurtle** [comment]
Okay, I think I figured it out.

- `:local` is only supposed to be used **when we download the model files directly to our git repo**. 
When Bumblebee downloads the files as found in [`cached_download/2`](https://github.com/elixir-nx/bumblebee/blob/57bdcced8d0df356093787ef38c02d76230b72a2/lib/bumblebee/huggingface/hub.ex#L52) (basically when calling `Bumblebee.load_model/2`), the files downloaded **are not the same as if we'd download the model to our git repo** - I'll call them `hashed files`. 

- `:ht` will cache downloads/`hashed files` in the same directory as set in `BUMBLEBLEE_CACHE_DIR`, as seen in `cached_download/2`. So we'll have an advantage if we download the model in the `Dockerfile`/put it in a volume and then make sure that our `BUMBLEBLEE_CACHE_DIR` points to it. That way, the model is fetched from this directory.

I had a lot of confusion on what the heck `:local` pertained to. So I hope this makes it more clear.

Now, I'm testing the container locally whilst downloading the model in the `Dockerfile` and now it seems to be working.
Here's the `Dockerfile`.

```dockerfile
# Find eligible builder and runner images on Docker Hub. We use Ubuntu/Debian
# instead of Alpine to avoid DNS resolution issues in production.
#
# https://hub.docker.com/r/hexpm/elixir/tags?page=1&name=ubuntu
# https://hub.docker.com/_/ubuntu?tab=tags
#
# This file is based on these images:
#
#   - https://hub.docker.com/r/hexpm/elixir/tags - for the build image
#   - https://hub.docker.com/_/debian?tab=tags&page=1&name=bullseye-20231009-slim - for the release image
#   - https://pkgs.org/ - resource for finding needed packages
#   - Ex: hexpm/elixir:1.15.7-erlang-26.0.2-debian-bullseye-20231009-slim
#
ARG ELIXIR_VERSION=1.15.7
ARG OTP_VERSION=26.0.2
ARG DEBIAN_VERSION=bullseye-20231009-slim

ARG BUILDER_IMAGE="hexpm/elixir:${ELIXIR_VERSION}-erlang-${OTP_VERSION}-debian-${DEBIAN_VERSION}"
ARG RUNNER_IMAGE="debian:${DEBIAN_VERSION}"

FROM ${BUILDER_IMAGE} as builder

# install build dependencies (and curl for EXLA)
RUN apt-get update -y && apt-get install -y build-essential git curl \
    && apt-get clean && rm -f /var/lib/apt/lists/*_*

# prepare build dir
WORKDIR /app

# install hex + rebar
RUN mix local.hex --force && \
    mix local.rebar --force

# set build ENV
ENV MIX_ENV="prod"
ENV BUMBLEBEE_CACHE_DIR="/app/.bumblebee/"
ENV BUMBLEBEE_OFFLINE="false"

# install mix dependencies
COPY mix.exs mix.lock ./
RUN mix deps.get --only $MIX_ENV
RUN mkdir config

# copy compile-time config files before we compile dependencies
# to ensure any relevant config change will trigger the dependencies
# to be re-compiled.
COPY config/config.exs config/${MIX_ENV}.exs config/
RUN mix deps.compile

COPY priv priv

COPY lib lib

COPY assets assets

COPY .bumblebee/ .bumblebee

# compile assets
RUN mix assets.deploy

# Compile the release
RUN mix compile

# IMPORTANT: This downloads the HuggingFace models from the `serving` function in the `lib/app/application.ex` file. 
# And copies to `.bumblebee`.
RUN mix run -e 'App.Application.load_models()' --no-start --no-halt; exit 0
COPY .bumblebee/ .bumblebee

# Changes to config/runtime.exs don't require recompiling the code
COPY config/runtime.exs config/

COPY rel rel
RUN mix release

# start a new build stage so that the final image will only contain
# the compiled release and other runtime necessities
FROM ${RUNNER_IMAGE}

RUN apt-get update -y && \
  apt-get install -y libstdc++6 openssl libncurses5 locales ca-certificates \
  && apt-get clean && rm -f /var/lib/apt/lists/*_*

# Set the locale
RUN sed -i '/en_US.UTF-8/s/^# //g' /etc/locale.gen && locale-gen

ENV LANG en_US.UTF-8
ENV LANGUAGE en_US:en
ENV LC_ALL en_US.UTF-8

WORKDIR "/app"
RUN chown nobody /app

# set runner ENV
ENV MIX_ENV="prod"

# Adding this so model can be downloaded
RUN mkdir -p /nonexistent

# Only copy the final release from the build stage
COPY --from=builder --chown=nobody:root /app/_build/${MIX_ENV}/rel/app ./
COPY --from=builder --chown=nobody:root /app/.bumblebee/ ./.bumblebee

USER nobody

# If using an environment that doesn't automatically reap zombie processes, it is
# advised to add an init process such as tini via `apt-get install`
# above and adding an entrypoint. See https://github.com/krallin/tini for details
# ENTRYPOINT ["/tini", "--"]

# Set the runtime ENV
ENV ECTO_IPV6="true"
ENV ERL_AFLAGS="-proto_dist inet6_tcp"
ENV BUMBLEBEE_CACHE_DIR="/app/.bumblebee/"
ENV BUMBLEBEE_OFFLINE="true"

CMD ["/app/bin/server"]
```

- so `BUMBLEBEE_OFFLINE` should **only be set to `true` *after* after the image is created and the app is compiled in the `Dockerfile`**. The reason it wasn't working before it's because `BUMBLEBEE_OFFLINE` was also being set during compilation. Somehow this messed up the process (though I'm not sure why).
- setting `BUMBLEBEE_CACHE_DIR` ending with `/` also helped, though I doubt it since `Path.join/1` is being used in `cache_downloads/2`.

Here's how my `application.ex` is looking.

```elixir
defmodule App.Application do
  # See https://hexdocs.pm/elixir/Application.html
  # for more information on OTP Applications
  @moduledoc false

  use Application

  @impl true
  def start(_type, _args) do
    children = [
      # Start the Telemetry supervisor
      AppWeb.Telemetry,
      # Start the PubSub system
      {Phoenix.PubSub, name: App.PubSub},
      # Nx serving for image classifier
      {Nx.Serving, serving: serving(), name: ImageClassifier},
      # Adding a supervisor
      {Task.Supervisor, name: App.TaskSupervisor},
      # Start the Endpoint (http/https)
      AppWeb.Endpoint
      # Start a worker by calling: App.Worker.start_link(arg)
      # {App.Worker, arg}
    ]

    # See https://hexdocs.pm/elixir/Supervisor.html
    # for other strategies and supported options
    opts = [strategy: :one_for_one, name: App.Supervisor]
    Supervisor.start_link(children, opts)
  end

  def load_models do
    # ResNet-50 -----
    {:ok, _} = Bumblebee.load_model({:hf, "microsoft/resnet-50"})
    {:ok, _} = Bumblebee.load_featurizer({:hf, "microsoft/resnet-50"})
  end

  def serving do
    # ResNet-50 -----
    {:ok, model_info} = Bumblebee.load_model({:hf, "microsoft/resnet-50"})
    {:ok, featurizer} = Bumblebee.load_featurizer({:hf, "microsoft/resnet-50"})

    Bumblebee.Vision.image_classification(model_info, featurizer,
      top_k: 1,
      compile: [batch_size: 10],
      defn_options: [compiler: EXLA],
      preallocate_params: true        # needed to run on `Fly.io`
    )

  end

  # Tell Phoenix to update the endpoint configuration
  # whenever the application is updated.
  @impl true
  def config_change(changed, _new, removed) do
    AppWeb.Endpoint.config_change(changed, removed)
    :ok
  end
end

```

The container now works locally without crashing and it seems to make `cache_download/2` fetch from the local file in `.bumblebee`.

See the video below.


https://github.com/dwyl/imgup/assets/17494745/3aa4ccac-d122-442c-8f83-b24a661e06e1


I've built the docker image with `--no-cache` on purpose. As you can see, if I restart the machine, it won't crash/manages to find the local model in `.bumblebee`.

The `metadata_filename` is the following piece of code that is found in `cache_download/2`:

```elixir
    url = "https://huggingface.co/api/models/microsoft/resnet-50/tree/main" |> :erlang.md5() |> Base.encode32(case: :lower, padding: false)
    metadata_filename = url <> ".json"

    dbg(metadata_filename)
```

This yields the `hashed file` `.json` file (which is `base32` encoded).

I digress. This should work for you too now 👌

**Comment by ndrean** [comment]
Yes, `:local` expects something else than what is downloaded.

But neither [Fly.io](https://fly.io/phoenix-files/speed-up-your-boot-times-with-this-one-dockerfile-trick/) nor [livebeats whisper](https://github.com/fly-apps/live_beats/blob/57f5c0f14277c6e476b9998c953e58d7deb0cd9d/Dockerfile#L61) use `exit 0` nor the `COPY` command that you use:

```elixir
RUN mix run -e 'App.Application.load_models()' --no-start --no-halt; exit 0
COPY .bumblebee/ .bumblebee
```

Still at the same point as 2 weeks ago: no clue why it fails this way.

**Comment by jeregrine** [comment]
```
COPY .bumblebee/ .bumblebee
```
Will copy from your context or local machine to the current image so doing it twice won't do anything. And docker is weird and finicky I'm sorry you're having these issues, but i find full paths work better than relative ones. 

If you do:
```
FROM ${builder}
RUN mix run -e 'App.Application.load_models()' --no-start --no-halt; exit 0
FROM ${runner}
COPY --from=builder /app/.bumblebee/ /app/.bumblebee
```

This example is also not great because you might create very large dockerfiles which make deploys slower. One thing we've been trying out is adding a volume and downloading the model on first boot to said volume.

```
unless File.exists?(model) do
   App.Application.load_models()'
end
```

Once your app deploys can you `fly ssh console` into it and verify that the `/app/.bumblebee` files are there or not. If they are then your configuration is wrong.

**Comment by LuchoTurtle** [comment]
Thanks for the feedback @jeregrine .

Yes, doing that will copy from my machine to the current image. It was actually working for a while but it stopped working after I've changed nothing. Weird stuff.

But yes, I'm aware this isn't the ideal solution - it creates gigantic image files, as you correctly stated. Having a volume is certainly the way to go and I'm currently exploring it. But I feel like my issue will still occur even with volumes. I'm testing stuff locally with `Docker` and I can see the model files being correctly downloaded, the env variables (`BUMBLEBEE_CACHE_DIR` and `BUMBLEBEE_OFFLINE`) are correctly set and *still* I get an error whilst loading the models that they are not found.

For example, in `application.ex`:

```elixir
 def load_models do
    # ResNet-50 -----
    {:ok, _} = Bumblebee.load_model({:hf, "microsoft/resnet-50"})
    {:ok, _} = Bumblebee.load_featurizer({:hf, "microsoft/resnet-50"})
  end

  def serving do

    dbg(System.get_env("BUMBLEBEE_CACHE_DIR"))
    dbg(System.get_env("BUMBLEBEE_OFFLINE"))

    # ResNet-50 -----
    {:ok, model_info} = Bumblebee.load_model({:hf, "microsoft/resnet-50"})
    {:ok, featurizer} = Bumblebee.load_featurizer({:hf, "microsoft/resnet-50"})

    Bumblebee.Vision.image_classification(model_info, featurizer,
      top_k: 1,
      compile: [batch_size: 10],
      defn_options: [compiler: EXLA],
      preallocate_params: true        # needed to run on `Fly.io`
    )

  end

```

where `serving/0` is used by `Nx` when the supervision tree is initiated.

If I run my `dockerfile`, I can clearly see that the files are being correctly downloaded and placed under `app/.bumblebee`.

<img width="904" alt="image" src="https://github.com/dwyl/imgup/assets/17494745/5691f03d-6d9b-4168-9d73-841b0dc4ed1b">


So, in theory, this should totally work. But it doesn't.

```sh
2023-11-14 15:06:33 [lib/app/application.ex:49: App.Application.serving/0]
2023-11-14 15:06:33 System.get_env("BUMBLEBEE_CACHE_DIR") #=> "/app/.bumblebee/"
2023-11-14 15:06:33 
2023-11-14 15:06:33 [lib/app/application.ex:50: App.Application.serving/0]
2023-11-14 15:06:33 System.get_env("BUMBLEBEE_OFFLINE") #=> "true"
2023-11-14 15:06:33 
2023-11-14 15:06:33 15:06:33.255 [info] TfrtCpuClient created.
2023-11-14 15:06:33 15:06:33.645 [notice] Application app exited: exited in: App.Application.start(:normal, [])
2023-11-14 15:06:33     ** (EXIT) an exception was raised:
2023-11-14 15:06:33         ** (MatchError) no match of right hand side value: {:error, "could not find file in local cache and outgoing traffic is disabled, url: https://huggingface.co/microsoft/resnet-50/resolve/main/preprocessor_config.json"}
2023-11-14 15:06:33             (app 0.1.0) lib/app/application.ex:54: App.Application.serving/0
2023-11-14 15:06:33             (app 0.1.0) lib/app/application.ex:16: App.Application.start/2
```


Regardless if the model is stored in a volume or not (I know that there are ephemeral storage considerations), I'm doing this on my computer and on a `Docker` instance. I'm at a loss at what I could be doing wrong 😅

**Comment by jeregrine** [comment]
Could you do a File.ls("/app/.bumblebee/") |> IO.inspect and see what you
get?

**Comment by LuchoTurtle** [comment]
@ndrean 
So, assuming you've created a volume called `models` on `fly.io` and you have this in your `fly.toml`.

```toml
[mounts]
  source = "models"
  destination = "/app/.bumblebee/"
```

At what stage do you download the models? Do you do it yourself manually? Do you run an external script that does this?

**Comment by ndrean** [comment]
This is my question! If you create a volume, can you ssh into it, even if no app is running (considering we are in the same region), or the simple "mounts" in the fly.toml will populate it?

**Comment by LuchoTurtle** [comment]
> Could you do a File.ls("/app/.bumblebee/") |> IO.inspect and see what you get?

I get the following: 

```elixir
2023-11-14 16:14:43 [lib/app/application.ex:49: App.Application.serving/0]
2023-11-14 16:14:43 System.get_env("BUMBLEBEE_CACHE_DIR") #=> "/app/.bumblebee/"
2023-11-14 16:14:43 
2023-11-14 16:14:43 [lib/app/application.ex:50: App.Application.serving/0]
2023-11-14 16:14:43 System.get_env("BUMBLEBEE_OFFLINE") #=> "true"
2023-11-14 16:14:43 
2023-11-14 16:14:43 {:ok, ["huggingface"]}
2023-11-14 16:14:43 [lib/app/application.ex:51: App.Application.serving/0]
2023-11-14 16:14:43 File.ls("/app/.bumblebee/") #=> {:ok, ["huggingface"]}
2023-11-14 16:14:43 |> IO.inspect() #=> {:ok, ["huggingface"]}
2023-11-14 16:14:43 
2023-11-14 16:14:43 16:14:43.822 [info] TfrtCpuClient created.
2023-11-14 16:14:44 16:14:44.224 [notice] Application app exited: exited in: App.Application.start(:normal, [])
2023-11-14 16:14:44     ** (EXIT) an exception was raised:
2023-11-14 16:14:44         ** (MatchError) no match of right hand side value: {:error, "could not find file in local ca
```

**Comment by LuchoTurtle** [comment]
I've added the following code to mimic [`cached_download/2`](https://github.com/elixir-nx/bumblebee/blob/57bdcced8d0df356093787ef38c02d76230b72a2/lib/bumblebee/huggingface/hub.ex#L52) and, as you can see, the filename matches and can be found inside `/app/.bumblebee/huggingface`.

```elixir

    url = "https://huggingface.co/api/models/microsoft/resnet-50/tree/main" |> :erlang.md5() |> Base.encode32(case: :lower, padding: false)
    metadata_filename = url <> ".json"
    dbg(metadata_filename)
    dbg(File.ls("/app/.bumblebee/huggingface") |> IO.inspect)
```

On startup, it yields...

```sh
2023-11-14T16:27:57.716 app[683d529c575228] mad [info] metadata_filename #=> "7p34k3zbgum6n3sspclx3dv3aq.json"

2023-11-14T16:27:57.717 app[683d529c575228] mad [info] {:ok,

2023-11-14T16:27:57.718 app[683d529c575228] mad [info] ["45jmafnchxcbm43dsoretzry4i.json",

2023-11-14T16:27:57.718 app[683d529c575228] mad [info] "7p34k3zbgum6n3sspclx3dv3aq.k4xsenbtguwtmuclmfdgum3enjuwosljkrbuc42govrhcudqlbde6ujc",

2023-11-14T16:27:57.718 app[683d529c575228] mad [info] "45jmafnchxcbm43dsoretzry4i.eiztamryhfrtsnzzgjstmnrymq3tgyzzheytqmrzmm4dqnbshe3tozjsmi4tanjthera",

2023-11-14T16:27:57.718 app[683d529c575228] mad [info] "7p34k3zbgum6n3sspclx3dv3aq.json",

2023-11-14T16:27:57.718 app[683d529c575228] mad [info] "6scgvbvxgc6kagvthh26fzl53a.ejtgmobrgyzwcmjtgiztgmztgezdmnzqgzsdmnbzmnstom3fmnsdontfgq2wimrugfrdimtegyzdgzdfme3ggnzsgm3dsmddmftgkmbxei",

2023-11-14T16:27:57.718 app[683d529c575228] mad [info] "6scgvbvxgc6kagvthh26fzl53a.json"]}
```

**Comment by LuchoTurtle** [comment]
> This is my question! If you create a volume, can you ssh into it, even if no app is running (considering we are in the same region), or the simple "mounts" in the fly.toml will populate it?

I don't think you can `ssh` into it without the app successfully running. I've tried and it kicks me out every time it attempts to restart the server (expected). As far as I can tell, `[mounts]` won't populate anything, it's just pointing where we want to the model to be. Though I don't quite understand how I'm meant to populate the volume in an automated manner lmao.

**Comment by ndrean** [comment]
> as you can see, the filename matches and can be found inside `/app/.bumblebee/huggingface`.

Does it matches all the files you have locally for this model? For me, no

**Comment by jeregrine** [comment]
My suggestion is to skip the dockerfile stuff. Do this

in your application startlink check iof the models exist, if not
download them to a volume. And it will be slow to boot once and you'll
never think about it again.

**Comment by LuchoTurtle** [comment]
Yeah, there's no use putting more effort on a deadend. So I'm downloading the model on the first boot up and then reusing it on subsequent restarts.

```elixir
  def start(_type, _args) do

	    # Checking if the models have been downloaded
	    models_folder_path = Path.join(System.get_env("BUMBLEBEE_CACHE_DIR"), "huggingface")
	    if not File.exists?(models_folder_path) or File.ls!(models_folder_path) == [] do
	      load_models()
	    end
	
		children = [
		...
	end

  def load_models do
    # ResNet-50 -----
    {:ok, _} = Bumblebee.load_model({:hf, "microsoft/resnet-50"})
    {:ok, _} = Bumblebee.load_featurizer({:hf, "microsoft/resnet-50"})
  end

  def serving do
    # ResNet-50 -----
    {:ok, model_info} = Bumblebee.load_model({:hf, "microsoft/resnet-50", offline: true})
    {:ok, featurizer} = Bumblebee.load_featurizer({:hf, "microsoft/resnet-50", offline: true})

    Bumblebee.Vision.image_classification(model_info, featurizer,
      top_k: 1,
      compile: [batch_size: 10],
      defn_options: [compiler: EXLA],
      preallocate_params: true        # needed to run on `Fly.io`
    )

  end
```

Instead of using `BUMBLEBEE_OFFLINE`, I'm passing the `:offline` option when loading the models, so they're fetched locally.

**Comment by LuchoTurtle** [comment]
> > as you can see, the filename matches and can be found inside `/app/.bumblebee/huggingface`.
> 
> Does it matches all the files you have locally for this model? For me, no

The filename from `cached_downloads/2` matches one of them (as expected). Though I see two files are missing when compared to the local one 💭 

Regardless, no use to keep trying with the `Dockerfile`-only approach anymore 😅

**Comment by ndrean** [comment]
So like me, some files are missing. Also the versions of Bumblebee and Nx are quite unstable.  I decided to fork your repo to run the same code as you did, but guess what, I can't even run `mix ph.server` because:

```bash
function EXLA.NIF.start_log_sink/1 is undefined (module EXLA.NIF is not available)
```


Starting to lose a bit my patience. But maybe more important,  the documentation isn't reliable? Take a look:  <https://github.com/elixir-nx/bumblebee/tree/main/examples/phoenix#tips>. What is really working?

**Comment by LuchoTurtle** [comment]
> So like me, some files are missing. Also the versions of Bumblebee and Nx are quite unstable. I decided to fork your repo to run the same code as you did, but guess what, I can't even run `mix ph.server` because:
> 
> ```shell
> function EXLA.NIF.start_log_sink/1 is undefined (module EXLA.NIF is not available)
> ```
> 
> Starting to lose a bit my patience. But maybe more important, the documentation isn't reliable? Take a look: [elixir-nx/bumblebee@`main`/examples/phoenix#tips](https://github.com/elixir-nx/bumblebee/tree/main/examples/phoenix?rgh-link-date=2023-11-14T21%3A04%3A44Z#tips). What is really working?

That's odd, I've never had that error happen to me. Does clearing out the deps and running `mix deps.get` again fix it? 
https://elixirforum.com/t/exla-nif-start-log-sink-1-issue-works-on-ubuntu-but-not-on-macbook-m2/58162/3 says it's a Linux-related issue, but it apparently has been solved.

From the link that you provided, from what I've gathered, I did follow the [`Configuring` Nx](https://github.com/elixir-nx/bumblebee/tree/main/examples/phoenix#configuring-nx) chapter and it seems to work the same as before. I can't really quantify it because I saw negligible impact after configuring it like so 🤷‍♂️

**Comment by ndrean** [comment]
Yes, I erased everything, mix.lock, deps, _build and restarted again. It works now.....

**Comment by ndrean** [comment]
Then for with or without EXLA, yes, there is a huge difference.
I did try a few months ago a very simple neural network with Axon, just to compute a linear regression by a simple gradient, and the difference was HUGE. Now, with what we are doing, no idea 😁

**Comment by ndrean** [comment]
I understand (?) that we are more of less loading the  coefficients of some kind of process that is used to define some Axon operations to build the neural network defined by the model. But what are we doing exactly, that I have absolutely no idea. But for sure, it is not rocket science 😁

**Comment by ndrean** [comment]
I recalled I made a Livebook last year to test Nx and Axon. The idea was to use a very simple example: linear regression. You start with the very well known  matrix formulas to compute the exact solution: whether by inverting a matrix, or using formulas.  The best fitting linear curve passing through a bunch of points is of course the linear curve which gives the minimal total euclidean dsitance between the curve and the pionts. You compare this to a smiple gradient descent, and since e are crazy, we can even use a NN! You "pompously" train your NN with your points, to build the coefficients of your NN. Then you can use it: given an input x, it finds an y. 
If you are interested just to see what is this about, I can paste somewhere the Livebook

**Comment by LuchoTurtle** [comment]
Do show! 

For my first NN I did something similar with Stochastic Gradient Descent -> https://github.com/LuchoTurtle/bike-sharing-patterns/blob/master/Your_first_neural_network.ipynb. 

Curious to see what you've done :)

**Comment by nelsonic** [comment]
@LuchoTurtle FYI: that link is `404` ... 🔗 🙈 
![image](https://github.com/dwyl/imgup/assets/194400/f10193a3-a91e-490a-a759-d80df1b564b6)
It it `public`? 🌎

**Comment by LuchoTurtle** [comment]
Should be now, thanks 👌

**Comment by ndrean** [comment]
Here: <https://github.com/ndrean/linear_regression_nx_axon>
But this is super modest work....
